File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.java
class Solution {
    public int findMin(int[] nums) {
        return findRecursive(nums, 0, nums.length - 1);
    }

    private static int findRecursive(int[] nums, int lb, int ub)   {
        if (lb == ub) {
            return nums[lb];
        }
        int mid = lb + (ub - lb) / 2;
        if (nums[mid] < nums[ub])    {
            return findRecursive(nums, lb, mid);
        } else {
            return findRecursive(nums, mid + 1, ub);
        }
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        arr_length = len(nums)
        product_array = [1] * arr_length

        prefix = 1      # to set first variable to not 0
        for i in range(arr_length): # to get first half of the multiplications
            product_array[i] *= prefix  # first set prefix and then change prefix so current index not present
            prefix *= nums[i]   # updated the prefix for the next go around. On last element it's lost

        suffix = 1
        for j in range(arr_length - 1, -1, -1):         # reversed so as to get the back half
            product_array[j] *= suffix  # first set suffix and then change suffix so current index not present
            suffix *= nums[j]   # updated the suffix for the next go around. On first element it's lost

        return product_array